hi-ml-histopathology/src/SSL/lightning_metrics.py

```diff
@@ -0,0 +1,385 @@
+#  ------------------------------------------------------------------------------------------
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  ------------------------------------------------------------------------------------------
+import math
+from enum import Enum, unique
+from typing import Any, Iterator, List, Tuple
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+import torchmetrics as metrics
+from torchmetrics.functional import accuracy, auc, auroc, precision_recall_curve, roc
+from torchmetrics import Metric
+from torch.nn import ModuleList
+
+
+AVERAGE_DICE_SUFFIX = "AverageAcrossStructures"
+TRAIN_PREFIX = "train/"
+VALIDATION_PREFIX = "val/"
+
+
+@unique
+class MetricType(Enum):
+    """
+    Contains the different metrics that are computed.
+    """
+    # Any result of loss computation, depending on what's configured in the model.
+    LOSS = "Loss"
+
+    # Classification metrics
+    CROSS_ENTROPY = "CrossEntropy"
+    # Classification accuracy assuming that posterior > 0.5 means predicted class 1
+    ACCURACY_AT_THRESHOLD_05 = "AccuracyAtThreshold05"
+    ACCURACY_AT_OPTIMAL_THRESHOLD = "AccuracyAtOptimalThreshold"
+    # Metrics for segmentation
+    DICE = "Dice"
+    HAUSDORFF_mm = "HausdorffDistance_millimeters"
+    MEAN_SURFACE_DIST_mm = "MeanSurfaceDistance_millimeters"
+    VOXEL_COUNT = "VoxelCount"
+    PROPORTION_FOREGROUND_VOXELS = "ProportionForegroundVoxels"
+
+    PATCH_CENTER = "PatchCenter"
+
+    AREA_UNDER_ROC_CURVE = "AreaUnderRocCurve"
+    AREA_UNDER_PR_CURVE = "AreaUnderPRCurve"
+    OPTIMAL_THRESHOLD = "OptimalThreshold"
+    FALSE_POSITIVE_RATE_AT_OPTIMAL_THRESHOLD = "FalsePositiveRateAtOptimalThreshold"
+    FALSE_NEGATIVE_RATE_AT_OPTIMAL_THRESHOLD = "FalseNegativeRateAtOptimalThreshold"
+
+    # Regression metrics
+    MEAN_ABSOLUTE_ERROR = "MeanAbsoluteError"
+    MEAN_SQUARED_ERROR = "MeanSquaredError"
+    EXPLAINED_VAR = "ExplainedVariance"
+
+    # Common metrics
+    SUBJECT_COUNT = "SubjectCount"
+    LEARNING_RATE = "LearningRate"
+
+
+def nanmean(values: torch.Tensor) -> torch.Tensor:
```

Comment:
nanmean is I think not used here? Also check if we really need those variants of MAE that count the number of observations